elastic/eui
# Add combined Jest+Cypress code coverage reports

## Summary

While working on #5261, I wanted to calculate combined coverage to make sure I wasn't losing any uncovered lines when converting the focus behavior tests to Cypress. This PR adds a Cypress code coverage plugin and then a `yarn combine-test-coverage` script that combines our Cypress+Jest output into a single HTML report. It additionally adds wiki documentation for how to access our HTML code coverage reports.

I recommend [following along by commit](https://github.com/elastic/eui/pull/5262/commits), as I added some misc. cleanup as part of this PR as well.

### Example

I cherry-picked #5261 into this branch locally to get an example of how updating a subsection of Jest tests to Cypress would affect final combined coverage. Coverage remained approximately the same (with a 1% increase from adding Cypress!) 

#### Before with just Jest coverage
<img width="757" alt="90.41%" src="https://user-images.githubusercontent.com/549407/137020500-122e5f7c-530c-406f-8269-f6f15de59975.png">

#### After

**Jest coverage:**
<img width="991" alt="82.19%" src="https://user-images.githubusercontent.com/549407/137020550-72e6a3ce-bfb1-415e-a485-6db3e2a821c2.png">

**Cypress coverage:**
<img width="1011" alt="67.12%" src="https://user-images.githubusercontent.com/549407/137020603-feccf35a-676a-49ac-a756-70ce63fba936.png">

**Combined coverage:**
<img width="1021" alt="91.78%" src="https://user-images.githubusercontent.com/549407/137020608-b816cd32-24b8-4e46-8cff-0568ba47b95b.png">

### QAing locally

- checkout this branch
- `yarn` to install new dependencies
- Run `yarn test-unit --coverage` to generate Jest coverage reports
- Run `yarn test-cypress` which will automatically generate Cypress coverage reports
- Run `yarn combine-test-coverage`, which should automatically open the main coverage report HTML page in your browser

### Resources

Reference links to the resources I used while I was setting this up:

- Cypress code coverage reports: [[1]](https://docs.cypress.io/guides/tooling/code-coverage), [[2]](https://github.com/cypress-io/code-coverage)
- Jest+Cypress combined coverage reports: [[1]](https://github.com/bahmutov/cypress-and-jest)
- Storyboook+Jest+Cypress combined coverage, if we're ever interested in that someday: [[1]](https://dev.to/penx/combining-storybook-cypress-and-jest-code-coverage-4pa5)

### Checklist

N/A, internal/dev-only change
